pkg/ebpf/bytecode/runtime/runtime_compilation_helpers.go

```diff
@@ -0,0 +1,172 @@
+// +build linux_bpf
+
+package runtime
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/ebpf/compiler"
+	"github.com/DataDog/datadog-agent/pkg/security/metrics"
+	"github.com/DataDog/datadog-agent/pkg/util/kernel"
+	"github.com/DataDog/datadog-agent/pkg/version"
+	"github.com/DataDog/datadog-go/statsd"
+)
+
+var (
+	defaultFlags = []string{
+		"-DCONFIG_64BIT",
+		"-D__BPF_TRACING__",
+		`-DKBUILD_MODNAME="ddsysprobe"`,
+		"-Wno-unused-value",
+		"-Wno-pointer-sign",
+		"-Wno-compare-distinct-pointer-types",
+		"-Wunused",
+		"-Wall",
+		"-Werror",
+	}
+)
+
+func ComputeFlagsAndHash(additionalFlags []string) ([]string, string) {
+	flags := make([]string, len(defaultFlags)+len(additionalFlags))
+	copy(flags, defaultFlags)
+	copy(flags[len(defaultFlags):], additionalFlags)
+
+	flagHash := hashFlags(flags)
+	return flags, flagHash
+}
+
+func hashFlags(flags []string) string {
+	h := sha256.New()
+	for _, f := range flags {
+		h.Write([]byte(f))
+	}
+	return fmt.Sprintf("%x", h.Sum(nil))
+}
+
+type RuntimeCompilationTelemetry struct {
+	compilationEnabled  bool
+	compilationResult   CompilationResult
+	compilationDuration time.Duration
+	headerFetchResult   kernel.HeaderFetchResult
+}
+
+func NewRuntimeCompilationTelemetry() RuntimeCompilationTelemetry {
+	return RuntimeCompilationTelemetry{
+		compilationEnabled: false,
+		compilationResult:  notAttempted,
+		headerFetchResult:  kernel.NotAttempted,
+	}
+}
+
+func (tm *RuntimeCompilationTelemetry) GetTelemetry() map[string]int64 {
+	stats := make(map[string]int64)
+	if tm.compilationEnabled {
+		stats["runtime_compilation_enabled"] = 1
+		stats["runtime_compilation_result"] = int64(tm.compilationResult)
+		stats["kernel_header_fetch_result"] = int64(tm.headerFetchResult)
+		stats["runtime_compilation_duration"] = tm.compilationDuration.Nanoseconds()
+	} else {
+		stats["runtime_compilation_enabled"] = 0
+	}
+	return stats
+}
+
+func (tm *RuntimeCompilationTelemetry) SendMetrics(client *statsd.Client) error {
+	tags := []string{fmt.Sprintf("version:%s", version.AgentVersion)}
+
+	var enabled float64 = 0
+	if tm.compilationEnabled {
+		enabled = 1
+	}
+	if err := client.Gauge(metrics.MetricRuntimeCompiledConstantsEnabled, enabled, tags, 1); err != nil {
+		return err
+	}
+
+	// if the runtime compilation is not enabled we return directly
+	if !tm.compilationEnabled {
+		return nil
+	}
+
+	if err := client.Gauge(metrics.MetricRuntimeCompiledConstantsCompilationResult, float64(tm.compilationResult), tags, 1); err != nil {
+		return err
+	}
+	if err := client.Gauge(metrics.MetricRuntimeCompiledConstantsCompilationDuration, float64(tm.compilationDuration), tags, 1); err != nil {
+		return err
+	}
+	return client.Gauge(metrics.MetricRuntimeCompiledConstantsHeaderFetchResult, float64(tm.headerFetchResult), tags, 1)
+}
```

Comment:
By submitting these metrics to statsd, it makes this metric customer-facing as opposed to internal. While there's nothing wrong with this approach, it does make it difficult to get any sort of meaningful summary of this telemetry, as you'd need to need query the metric on a customer-by-customer basis and manually aggregate the data yourself.

To get around this issue for the other runtime compiled assets, the network tracer takes this telemetry and attaches it to the network flow data which it gathers. Then, when our backend processes the network flow data, it  also aggregates and emits the runtime compilation telemetry gathered (if you're interested, feel free to ping me on Slack & I can show you where/how this is done). This allows us to create internal dashboards which track the success/failure of runtime compilation on an asset-by-asset basis across all hosts which have the network tracer enabled (including our customer's!).

If you'd like the telemetry from the `RuntimeCompilationConstantFetcher` asset to be included in this aggregation, all you need to do is add it here:
https://github.com/DataDog/datadog-agent/blob/7fecda3dc6b785e286e74aae83691e397e591fc8/pkg/network/tracer/tracer.go#L371-L378

Reply:
FWIW, you could use both approaches if you want!